### ISSUE TITLE
PictureRecorder needs a dispose method

### ISSUE THREAD

**Issue by jezell** [flutter/flutter#146278]
### Use case

While investigating #145220 we discovered that if there are errors encountered during rendering, the renderer leaks picture recorders and will eventually run the app out of memory. This is something that is addressable, but not without PictureRecorder implementing dispose. It holds on to native resources (for example SkPictureRecorder in CanvasKit) yet you can't properly dispose it. The assumption of the class is that you will always call endRecording, but that may not be the case when handling exceptions and it certainly isn't the case when endRecording throws due to OOM conditions, which may be temporary.

### Proposal

Implement dispose on PictureRecorder.

**Comment by goderbauer** [comment]
Tagging this with engine since the PictureRecorder is part of the dart:ui API.

**Comment by jonahwilliams** [comment]
I don't think we have enough information to conclude that we should add this API. Tagging web team to investigate their implementation to confirm if it leaks or not.

**Comment by jezell** [comment]
@yjbanov I believe embind already uses the FinalizationRegistry for everything, but there is a big warning right on the page that you really need should manually call delete. The finalization registry may or may not actually dispose the objects. In the picture recorder case you could be leaking 60 frames per second of picture recorders and go OOM in canvaskit heap rather fast if you depend on GC to eventually clean them up.

<img width="743" alt="Screenshot 2024-05-09 at 11 37 11 AM" src="https://github.com/flutter/flutter/assets/22264/44909bfe-a2ff-4ef9-b3d4-71a020dbf05e">
